### ISSUE TITLE
BUG BIND DNS Server listening to all IP addresses when 4 specified in settings.

### ISSUE THREAD

**Issue by chris001** [virtualmin/virtualmin-gpl#441]
The only 4 addresses specified in BIND settings to listen on are: 
1. one public IPv4 address, 
2. one public IPv6 address, 
3. IPv4 `localhost`, and 
4. IPv6 `localhost`.

However BIND is listening to all 9 of the Virtualmin server's IPv4 and IPv6 addresses, which blocks another local DNS service (`dnsmasq`) from listening on its own private non-routable IP address `10.x.x.x`, which is required by `LXD` to assign IP addresses to its containers.

**Comment by chris001** [comment]
This is separate, however, related.
Seems I found an undocumented "feature" or unexpected behavior of BIND9 that was causing BIND9 to ignore my directive to not listen on the IP range used by `lxd`.
You have to put the ignore IP range first, before the listen on IP addresses, otherwise BIND will listen on the IP address or range you wanted it to ignore.
`listen-on { my.ip.v4.address; localhost; !10.0.0.0/8; }` fails.
`listen-on { !10.0.0.0/8; my.ip.v4.address; localhost; }` works.. so that BIND9 does not listen on any `10.x.x.x` address needed by `lxd` `dnsmasq`.

**Comment by jcameron** [comment]
Is there anything we could do in Virtualmin to avoid this?

**Comment by chris001** [comment]
> Is there anything we could do in Virtualmin to avoid this?

Yes!

![bind addresses and topology](https://user-images.githubusercontent.com/259416/188523117-de732a00-f1c0-4a62-a096-c50a93a4b4ae.png)

On the `BIND DNS Server` settings page `Addresses and Topology`, in the field `Addresses` ...

1. Extract the IPv4 addresses (and host names that resolve to an IPv4 address, e.g. `localhost`), sort the "not" ones first (they start with `!`), and output them to the `listen-on` directive.
2. Extract the IPv6 addresses (and host names that resolve to an IPv6 address e.g. `localhost`), sort the "not" ones first (they start with `!`), and output them to the `listen-on-v6` directive.
3. Optional advanced, add a button in UI and CLI to enable [`DNS over TLS` (secure encrypted DNS for which BIND listens on the IETF standard port `853`)](https://datatracker.ietf.org/doc/rfc7858/), when on/off, add or uncomment/comment out, these directives (`tls`, `listen-on tls` and `listen-on-v6 tls`):

Example `/etc/bind/named.conf.options` :
```
tls local-tls {
   key-file  "/etc/letsencrypt/live/my.bind.server.tld/privkey.pem";
   cert-file "/etc/letsencrypt/live/my.bind.server.tld/fullchain.pem";
};

options {
// ....................other lines...
     listen-on { !not.this.ipv4.address; my.public.ipv4.address; localhost; };
     listen-on-v6 { !not.this.ipv6.address; my.public.IPv6.address; localhost; };

       listen-on tls local-tls { !not.this.ipv4.address; my.public.IPv4.address; localhost; };
	listen-on-v6 tls local-tls { !not.this.ipv6.address; my.public.IPv6.address; localhost; };
}
```

**Comment by chris001** [comment]
Is it possible for BIND9 to be behaving differently than the settings?  It's as if BIND9 is doing a `listen-on { any; }` and `listen-on-v6 { any; }' to listen on ALL the IP addresses on the system including local non-routable addresses, then for it to apply the settings from the `named.local.options` file, for `listen-on`, and `listen-on-v6` (in particular, which IP addresses NOT to listen on, to un-listen on those addresses).  End result, it's listening on way too many IP addresses, which prevents `dnsmasq` from listening on the (local non-routable) IP addresses for the bridge interface it expects to listen on?

**Comment by iliajie** [comment]
> Is it possible for BIND9 to be behaving differently than the settings? It's as if BIND9 is doing a `listen-on { any; }` and `listen-on-v6 { any; }' to listen on ALL the IP addresses on the system including local non-routable addresses`

By default BIND configured to listen on any address from all available interfaces using `listen-on { any; };` option. If you want only specific address(es) you can configure them on `Servers ⇾ BIND DNS Server: Addresses and Topology` page using `Ports and addresses to listen on` text box, like `100.100.100.222 200.200.200.111`.

However, I agree that it may be confusing at first that adding only `!111.222.111.222` to the field, will leave BIND listening only on `127.0.0.1`. But this is how BIND works and I don't think we should do anything here.

I think what you could do in your case @chris001 is to specify address to exclude and add `any` at the end on the line, like `!192.168.10.2 !192.168.10.3 !192.168.10.4 any`, which will leave BIND listening on all available IPv4 addresses except mentioned. 

Jamie, also I think we need to improve IPv6 address support. As right now it's possible to post `fe80::11:32ff:fe26:5` as address, and have it placed inside of `listen-on` rather than `listen-on-v6` directive.

**Comment by jcameron** [comment]
Yeah, we don't support `listen-on-v6` at all yet! I should fix that ..

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
